### PR TITLE
Revive sys-cluster/slurm

### PR DESCRIPTION
**Pull request by heroxbd**
Revive sys-cluster/slurm that is tree-cleaned for multiple CVEs earlier this year.

Please check all the boxes that apply:

- [X] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [X] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [X] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [X] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

File: sys-cluster/slurm/files/slurmd.initd [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	fi
+	# Make sure folders exists
+	. /etc/slurm/slurm.conf
+	folders=(
+		${SlurmdSpoolDir}
+		$(dirname ${SlurmctldLogFile})
+		$(dirname ${SlurmctldPidFile})
+		$(dirname ${SlurmdLogFile})
+		$(dirname ${SlurmdPidFile})
+		${SlurmdSpoolDir}
```
Comment (orlitzky) [inline sys-cluster/slurm/files/slurmd.initd:46]: `SlurmdSpoolDir` appears twice :)

More importantly, there is a potential for an exploit here. Imagine the following scenario as root:

```
# chown slurm:slurm /foo
# chown slurm:slurm /foo/bar
```

When the two commands are run in that order, there is a small window where the slurm user can replace `/foo/bar` with a link. The `checkpath` utility tries hard to guard against that, but there's still a race condition.

In my opinion it's simpler to pick reasonable default values for these paths, and make them work out-of-the-box by setting permissions in the ebuild. When everything works, users will be much less likely to mess with them, and then you don't have to worry about "fixing" the permissions in the first place. Up to you, but I had to mention it!

Comment (heroxbd) [reply sys-cluster/slurm/files/slurmd.initd:46]: You are right Michael, the default `slurm.conf.example` gives:

```
SlurmctldPidFile=/run/slurm/slurmctld.pid
SlurmdPidFile=/run/slurm/slurmd.pid
SlurmdSpoolDir=/var/spool/slurm/slurmd
SlurmctldLogFile=/var/log/slurmctld.log
SlurmdLogFile=/var/log/slurmd.log
```

The directories `/var/log` exists, `/var/spool/slurm` is created by the ebuild, and `/run/slurm` is managed by tmpfiles.  If users change the default, I will let the script exit if the needed directories does not exist.

Comment (heroxbd) [reply sys-cluster/slurm/files/slurmd.initd:46]: Hmm, after reading the manpage of `openrc-run(8)`, `checkpath` seems to guard against the symlink attack, although I don't know the details underneath.   I shall fix the duplication for the moment and keep this security concern in mind.

File: sys-cluster/slurm/slurm-24.05.3.ebuild [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+		$(use_with ucx) \
+		$(use_with yaml) \
+		$(use_enable static-libs static) \
+		$(use_enable slurmrestd) \
+		$(use_enable multiple-slurmd)
+
+	# --htmldir does not seems to propagate... Documentations are installed
+	# in /usr/share/doc/slurm-2.3.0/html
+	# instead of /usr/share/doc/slurm-2.3.0.2/html
+	sed \
```
[the remark this answers is not in the document]
Comment (heroxbd) [reply sys-cluster/slurm/slurm-24.05.3.ebuild:160]: This is no longer needed.

File: sys-cluster/slurm/slurm-24.05.3.ebuild [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+		CPPFLAGS="-I/opt/cuda/include"
+		--sysconfdir="${EPREFIX}/etc/${PN}"
+		--with-hwloc="${EPREFIX}/usr"
+		--htmldir="${EPREFIX}/usr/share/doc/${PF}"
+	)
+	use pam && myconf+=( --with-pam_dir=$(getpam_mod_dir) )
+	use mysql || myconf+=( --without-mysql_config )
+	econf "${myconf[@]}" \
+		$(use_enable debug) \
+		$(use_with lua) \
```
[the remark this answers is not in the document]
Comment (heroxbd) [reply sys-cluster/slurm/slurm-24.05.3.ebuild:143]: Good idea!

File: sys-cluster/slurm/slurm-24.05.3.ebuild [hunk 4]
```diff
@@ -0,0 +1,278 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
```
[the remark this answers is not in the document]
Comment (heroxbd) [reply sys-cluster/slurm/slurm-24.05.3.ebuild:4]: Right away.

File: sys-cluster/slurm/files/slurmd.initd [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+checkconfig() {
+	if [ ! -e "/etc/slurm/slurm.conf" ]; then
+		eerror "Missing config /etc/slurm/slurm.conf"
+		eerror "Customize sample one or generate new by configurator.html"
+		eerror "in slurm doc directory"
+		return 1
+	fi
+	# Make sure folders exists
+	. /etc/slurm/slurm.conf
+	folders=(
```
[the remark this answers is not in the document]
Comment (heroxbd) [reply sys-cluster/slurm/files/slurmd.initd]: fixed the bashisms.

File: sys-cluster/slurm/files/slurmd.initd (hunk 1 above)
Comment (orlitzky) [reply sys-cluster/slurm/files/slurmd.initd]: Symlinks should be safe but there's still a race condition in the hardlink detection. In pseudo-code, checkpath does,

```
if this inode has more than one link {
  error
}
else {
  change the permissions
}
```

but it's possible for the number of links to change between the first and last line. On linux these attacks are preventable with the `fs.protected_hardlinks` sysctl so they're not a huge problem. But I've found that refactoring to avoiding them tends to make the code simpler anyway.

Comment (heroxbd) [reply sys-cluster/slurm/files/slurmd.initd]: Interesting!  I read about `protected_hardlinks` and got the same impression.  Noted.

Thanks Michael for the enlightenment.

**Pull request closed without merging**